### ISSUE TITLE
All topics and users deleted

### ISSUE THREAD

**Issue by qbast** [banzaicloud/kafka-operator#186]
**Describe the bug**

In last week, two times I found all topics and users deleted from the cluster. The cluster was also in inoperable state (broker unable to start), I will create a separate bug for that.

**Steps to reproduce the issue:**
I do not know what triggered this problem.

**Additional context**
Using operator 0.2.3 installed via helm in GKE (1.14.7-gke.14) . KafkaCluster named 'kafka' is installed in 'infra' namespace, while KafkaTopic and KafkaUser objects are created in 'backend' namespace.

At some point (1-2 days after creation of the cluster), garbage collector suddenly deletes all topics and users. From the GKE audit log:

```
A 2019-11-05T23:01:23.950598Z Kubernetes Apiservice Requests update backend:audit-events system:serviceaccount:infra:kafka-operator-operator {"@type":"type.googleapis.com/google.cloud.audit.AuditLog","authenticationInfo":{"principalEmail":"system:serviceaccount:infra:kafka-operator-operator"},"authorizationInfo":[{"granted":true,"permission":"io.banzaicloud.kafka.v1alpha1.kafkatopics.update","resource":"kafka.banzaicloud.io/v1alpha1/name… Kubernetes Apiservice Requests update backend:audit-events system:serviceaccount:infra:kafka-operator-operator 
A 2019-11-05T23:01:23.959102Z Kubernetes Apiservice Requests update backend:audit-events system:serviceaccount:infra:kafka-operator-operator {"@type":"type.googleapis.com/google.cloud.audit.AuditLog","authenticationInfo":{"principalEmail":"system:serviceaccount:infra:kafka-operator-operator"},"authorizationInfo":[{"granted":true,"permission":"io.banzaicloud.kafka.v1alpha1.kafkatopics.update","resource":"kafka.banzaicloud.io/v1alpha1/name… Kubernetes Apiservice Requests update backend:audit-events system:serviceaccount:infra:kafka-operator-operator 
A 2019-11-05T23:01:23.967825Z Kubernetes Apiservice Requests update namespaces:kafkatopics:status system:serviceaccount:infra:kafka-operator-operator {"@type":"type.googleapis.com/google.cloud.audit.AuditLog","authenticationInfo":{"principalEmail":"system:serviceaccount:infra:kafka-operator-operator"},"authorizationInfo":[{"granted":true,"permission":"io.banzaicloud.kafka.v1alpha1.kafkatopics.status.update","resource":"kafka.banzaicloud.io/v1alph… Kubernetes Apiservice Requests update namespaces:kafkatopics:status system:serviceaccount:infra:kafka-operator-operator 
A 2019-11-05T23:01:24.167920Z Kubernetes Apiservice Requests update backend:twilio-api-delivery system:serviceaccount:infra:kafka-operator-operator {"@type":"type.googleapis.com/google.cloud.audit.AuditLog","authenticationInfo":{"principalEmail":"system:serviceaccount:infra:kafka-operator-operator"},"authorizationInfo":[{"granted":true,"permission":"io.banzaicloud.kafka.v1alpha1.kafkatopics.update","resource":"kafka.banzaicloud.io/v1alpha1/name… Kubernetes Apiservice Requests update backend:twilio-api-delivery system:serviceaccount:infra:kafka-operator-operator 
A 2019-11-05T23:01:24.197560Z Kubernetes Apiservice Requests update backend:twilio-api-delivery system:serviceaccount:infra:kafka-operator-operator {"@type":"type.googleapis.com/google.cloud.audit.AuditLog","authenticationInfo":{"principalEmail":"system:serviceaccount:infra:kafka-operator-operator"},"authorizationInfo":[{"granted":true,"permission":"io.banzaicloud.kafka.v1alpha1.kafkatopics.update","resource":"kafka.banzaicloud.io/v1alpha1/name… Kubernetes Apiservice Requests update backend:twilio-api-delivery system:serviceaccount:infra:kafka-operator-operator 
A 2019-11-05T23:01:24.216953Z Kubernetes Apiservice Requests update namespaces:kafkatopics:status system:serviceaccount:infra:kafka-operator-operator {"@type":"type.googleapis.com/google.cloud.audit.AuditLog","authenticationInfo":{"principalEmail":"system:serviceaccount:infra:kafka-operator-operator"},"authorizationInfo":[{"granted":true,"permission":"io.banzaicloud.kafka.v1alpha1.kafkatopics.status.update","resource":"kafka.banzaicloud.io/v1alph… Kubernetes Apiservice Requests update namespaces:kafkatopics:status system:serviceaccount:infra:kafka-operator-operator 
A 2019-11-05T23:01:36.007443Z Kubernetes Apiservice Requests update kafkatopics.kafka.banzaicloud.io system:serviceaccount:cert-manager:cert-manager-cainjector {"@type":"type.googleapis.com/google.cloud.audit.AuditLog","authenticationInfo":{"principalEmail":"system:serviceaccount:cert-manager:cert-manager-cainjector"},"authorizationInfo":[{"granted":true,"permission":"io.k8s.apiextensions.v1beta1.customresourcedefinitions.update","resource":"apiextensions.… Kubernetes Apiservice Requests update kafkatopics.kafka.banzaicloud.io system:serviceaccount:cert-manager:cert-manager-cainjector 
A 2019-11-05T23:03:39.901775Z Kubernetes Apiservice Requests delete backend:audit-events system:serviceaccount:kube-system:generic-garbage-collector {"@type":"type.googleapis.com/google.cloud.audit.AuditLog","authenticationInfo":{"principalEmail":"system:serviceaccount:kube-system:generic-garbage-collector"},"authorizationInfo":[{"granted":true,"permission":"io.banzaicloud.kafka.v1alpha1.kafkatopics.delete","resource":"kafka.banzaicloud.io/v1alp… Kubernetes Apiservice Requests delete backend:audit-events system:serviceaccount:kube-system:generic-garbage-collector 
A 2019-11-05T23:03:40.095352Z Kubernetes Apiservice Requests delete backend:twilio-api-delivery system:serviceaccount:kube-system:generic-garbage-collector {"@type":"type.googleapis.com/google.cloud.audit.AuditLog","authenticationInfo":{"principalEmail":"system:serviceaccount:kube-system:generic-garbage-collector"},"authorizationInfo":[{"granted":true,"permission":"io.banzaicloud.kafka.v1alpha1.kafkatopics.delete","resource":"kafka.banzaicloud.io/v1alp… Kubernetes Apiservice Requests delete backend:twilio-api-delivery system:serviceaccount:kube-system:generic-garbage-collector 
A 2019-11-05T23:03:41.580060Z Kubernetes Apiservice Requests delete infra:kafka-cruise-control-topic system:serviceaccount:kube-system:generic-garbage-collector {"@type":"type.googleapis.com/google.cloud.audit.AuditLog","authenticationInfo":{"principalEmail":"system:serviceaccount:kube-system:generic-garbage-collector"},"authorizationInfo":[{"granted":true,"permission":"io.banzaicloud.kafka.v1alpha1.kafkatopics.delete","resource":"kafka.banzaicloud.io/v1alp… Kubernetes Apiservice Requests delete infra:kafka-cruise-control-topic system:serviceaccount:kube-system:generic-garbage-collector 
```

The operator explicitly allows KafkaTopic/KafkaUser object to reference cluster in another namespace, but is it actually supported by kubernetes? According to documentation ( https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/ ) ownerReferences only work in the same namespace (and for cluster-wide resources).  There is also an open bug ( https://github.com/kubernetes/kubernetes/issues/65200 ) that makes cross-namespace ownerReferences behave in non-deterministic way. For example deleting all owned resources on restart of controller which might be happening in my case.

**Comment by tinyzimmer** [comment]
I haven't tested this personally on K8s 1.14.x, but on 1.15.x I was able to create a cluster in a `kafka` namespace and topics/users in the `default` namespace without issue. So I had assumed this functionality was solid. But looking at the doc you posted you are probably right that owner references should not be crossing namespaces.

A branch or two ago I switched it to placing labels on children `KafkaTopic` and `KafkaUser` resources, and using those to find resources to cleanup on a `KafkaCluster` deletion. I think it might be possible to take those owner references out now as they are redundant and it might take care of this.

**Comment by qbast** [comment]
You can create such cluster  and it will run without issue - for a time. The kubernetes bug I linked mentions that cross-namespace references should be immediately treated as missing. Instead it kind of works, until garbage collector controller restarts and it happens to process KafkaTopic/KafkaUser before processing KafkaCluster (so cluster's UID is not yet cached) - so we have non-derministic behaviour triggered by a race condition after controller restart.

**Comment by tinyzimmer** [comment]
Tasty. Well yea previous possible solution still stands. I can probably get this taken care of in the next day or two 😄. The owner references between users and their certificates will remain (as they are always in the same namespace), but the one between the cluster and topics/users is not really necessary anymore.

**Comment by tinyzimmer** [comment]
Update, I am going to take care of this on the PR I currently have open.

Someone might correct me on this, but if I understand correctly, that branch will split into two release tracks. There will be a point release on 0.7.x that contains the bugfixes and remains on `cert-manager` v0.10.x, and a new 0.8.x that runs on `cert-manager` 0.11.x. The update process between the two is a small PITA so it made sense to try to support both for a bit.

**Comment by baluchicken** [comment]
The PR #180  @tinyzimmer mentioned is in on both branches(master and 0.7.x). For multiple cert manager support we are going to support two releases, like you just said.

**Comment by baluchicken** [comment]
We just released Operator version 0.7.2 and 0.8.0 which contains the fix for this.
